### PR TITLE
feat(editors/subscription) Add supervision indication

### LINKED ISSUE #942: Add `LGOS` and `LSVS` acc. to IEC 61850-6 and IEC 61850-7-1
As a user of OpenSCD I want the tool to handle the `LGOS` and `LSVS` behavior acc. to the IEC 61850-6 2007B4. that is

**Background**
It is very easy for the tester of IEC 61850 communication to test if a GOOSE or Sampled Value Stream is published on the network as expected, as both are multicast messages that can be received on any port and compared to its setting in the SCD file. Testing of supervision is not so easy. With edition 1 it was a bad trial and error process until the IED did use the signal and the tester could see either measured values or some binary information triggering something....for that reason the logical node classes were introduced with edition 2: `LGOS` and `LSVS`. They allow per GOOSE and Sampled Value instance to supervise the subscription status. 

As all logical nodes `LGOS` and `LSVS` are allocated in the data model. In this case, in the receiving IED. With it the question arise who is responsible to put the data into the data model. With Ed2 this question was not answered fully and this lead to the discussion in [Tissue 1401](https://iec61850.tissue-db.com/tissue/1401). With the newest part of the 61850-6 this is clarified to an extent that we can be confident enough to implement it. See Table 11 in IEC 61850-6 for `SupSubscription` element for the wording of the -6.

**Requirements**
- only relevant for >= Ed2(2007B)
- Triggered on each newly subscribed GOOSE/SampledValue: no information from the GOOSE/SampledValue has been subscribed in the subscriber IED yet (query `ExtRef[iedName="sendingIED"][srcCBName="name of GOOSE/SampledValue"]` in the subscribing IED returns null)
- first try to add the control block reference to an existing but not used `LGOS`/`LSVS`
- if this is not possible, try to add new `LGOS`/`LSVS` to the subscribing IED
- warn the user if new instance of `LN[lnClass="LSVS"]` exceed the `maxSv` number

>NOTE: There are two possible scenarios that can happen and the requirements differ a bit: 1. There are already pre-defined `LGOS`/`LSVS` in the data model, and we need to instantiate the `DOI[name="GoCBRef"]>DAI[name="setSrcRef"]` or `DOI[name="SvCBRef"]>DAI[name="setSrcRef"]`, respectively. 2. There are no instantiated `LGOS`/`LSVS` and we need to instantiated new once in order to set the `DOI[name="GoCBRef"]>DAI[name="setSrcRef"]` or `DOI[name="SvCBRef"]>DAI[name="setSrcRef"]`, respectively. 

**Configure existing `LGOS`/`LSVS`**
1. check if the subscribing IED does have an unused instance of `LGOS`/`LSVS`
- get the first element `LN[lnClass="LGOS"]`/`LN[lnClass="LSVS"]` that does not have
- `DOI[name="GoCBRef"]>DAI[name="setSrcRef"]`/`DOI[name="SvCBRef"]>DAI[name="setSrcRef"]`
2. Add the GOOSE control block reference / the Sampled Values control block reference (e.g. `AA1E1Q01MUMod2_MU1/LLN0.MSVCB01`) as a value to the `DAI[name="setSrcRef"]`
>NOTE: the control block reference is a pointer to the `GSEControl`/`SampledValueControl` element that is structured as follows IEDNameLDInst/LLN0.CBName. (the `name` attribute of the receiving IED followed by the `inst` attribute of the parent LDevice follow by `/` follow by `LLN0` followed by `.` followed by the `name` attribute of the GOOSE/SampledValue control block)
3. Add a child `<Private type="OpenSCD.create"/>` to the `LGOS/LSVS` element

**First instantiate than configure `LGOS`/`LSVS`**
1. check all the pre-condition for adding LGOS/LSVS further down
2. if all conditions are met, add a new `LN` element into the logical device `LDevice` where all the other `LN[lnClass="LGOS"]`/`LN[lnClass="LSVS"]`
>NOTE: One pre-condition is that at least one such logical node exist already in the data model

**Pre-Condition to add new LN instance of LGOS/LSVS**
1. Check if `LGOS` instance can be added to the subscriber IED (all checks are related to the subscriber IED!)
- Check in the `Services` element - child of `IED` or `AccessPoint` if `Services>ClientServices>SubSubscription.maxGo` is bigger than 0 
- Check if the subscriber IED has instantiated logical node `LN[lnClass="LGOS"]`
- Check if the `lnType` attribute of instantiated logical node has a representation in the data type templates section `DataTypeTeplates>LNodeType[id="${lnType}"]`
- check if the number of the already instantiated `LGOS` is smaller than the allowed subscription for the subscriber IED: `querySelectorAll(LN[lnClass="LGOS"]).lenght < Services>ClientServices>SubSubscription.maxGo`
- check if the IED allows instantiating values `DataTypeTemplates>LNodeType[lnClass="LGOS"]>DO[name="GoCBRef"]>DA[name="setSrcRef"]` has attributes `valKind=Conf/RO` and `valImport=true`. 
> NOTE: `valImport` defines if instantiated data (instantiated by SCT) will change the data model when re-importing into the ICT. This is the case here. We want instantiate the `setSrcRef` data attribute and wan the IED to re-import and use it

2. Check if `LSVS` instance can be added to the subscriber IED (all checks are related to the subscriber IED!)
- Check in the `Services` element - child of `IED` or `AccessPoint` if `Services>ClientServices>SubSubscription.maxSv` is bigger than 0 
- Check if the subscriber IED has instantiated logical node `LN[lnClass="LSVS"]`
- Check if the `lnType` attribute of instantiated logical node has a representation in the data type templates section `DataTypeTeplates>LNodeType[id="${lnType}"]`
- check if the number of the already instantiated `LSVS` is smaller than the allowed subscription for the subscriber IED: `querySelectorAll(LN[lnClass="LSVS"]).lenght < Services>ClientServices>SubSubscription.maxSv`
- check if the IED allows instantiating values `DataTypeTemplates>LNodeType[lnClass="LSVS"]>DO[name="SvCBRef"]>DA[name="setSrcRef"]` has attributes `valKind=Conf/RO` and `valImport=true`. 
> NOTE: `valImport` defines if instantiated data (instantiated by SCT) will change the data model when re-importing into the ICT. This is the case here. We want instantiate the `setSrcRef` data attribute and wan the IED to re-import and use it

>NOTE: see the IEC 61850-7-1 Annex G: LGOS/LSVS engineering. This a very nice overview of the several checks we need to implement to make sure we are conflicting with the ICT (IEC configuration tool)

**Additional information**
- see the [tissue database](https://iec61850.tissue-db.com/default.mspx) and search for `LGOS`
- see IEC 61850-6 Table 11
- see IEC 61850-7-1 Annex G

### LINKED ISSUE #1037: Make LGOS and LSVS accessible in subscription editors
**Is your feature request related to a problem? Please describe.**

When #942 is complete, supervision of subscriptions will be provided.
But it will not be visible to the user.

As a user, I care about this because I want to know:

* Where is this subscription supervised?
* When browsing the SCL object model online, which LSVS or LGOS logical node is associated with the communications of this FCDA or Control Element.

**Describe the solution you'd like**

I would like to see the path to the LN and the LN name and a small icon to indicate supervision, perhaps `monitor_heart`

Perhaps something like one of these for the later binding plugin:

image

Perhaps only the LN information could be shown with the fully qualified path being a tooltip.

Another option might be to show only the `monitor_heart` icon with the tooltip being the path and the LGOS name which would reduce clutter. This would mean "this subscription is supervised" and the user would need to hover to see the detail. The icon could be a button which opened the Logical Node editor for the user to access additional details.

In the Subscriber(GOOSE) and Subscriber(SMV) plugins, perhaps it could be on the right.

image

image

### PR DESCRIPTION
**Pull request by danyill**
Closes #1037 

Possibly helpful for functional reviewing?

### CODE REVIEW

File: src/editors/subscription/foundation.ts [hunk 1]
```diff
@@ -343,6 +344,56 @@ export function findOrCreateAvailableLNInst(
...
+    ?.querySelectorAll('LN0 > Inputs, LN > Inputs')
+    .forEach(inputs => {
+      Array.from(dataSet!.querySelectorAll('FCDA')).some(fcda => {
+        const anExtRef = getExtRef(inputs, fcda, publishedControlBlock);
+        if (anExtRef) {
+          extRef = anExtRef
+          return true;
+        }
+      });
+    });
```
Comment (danyill) [inline src/editors/subscription/foundation.ts:366]: This is not very efficient, I must improve the way the iteration is done.

**Comment by danyill** [comment]
OK, I think it's possible that if the tests are working then this will pass. I've got issues running the tests locally at the moments (a lot of test failures with timeouts and I do not know why).

1. I've augmented the message binding test files with LGOS and LSVS supervision nodes.
2. I've done what I can with the unit tests.
3. The logic of the `getFirstSubscribedExtRef` function is not very clean and I will be interested in any suggestions for improvement.

I would be most grateful for a review :smile:

**Comment by danyill** [comment]
While watching a [:rocket: launch](https://www.rocketlabusa.com/live-stream/) I realised I was returning the identity of Val not the LN, so I have corrected.

**Comment by JakobVogelsang** [comment]
It looks all good to me @danyill. Nice job:) I would wait for #942 to go in? Is that ok with you. I think there would be many merge conflicts otherwise.

**Comment by danyill** [comment]
Sure, thanks for taking a look

I targeted [openscd:942-subscriber-add-lgos-and-lsvs](https://github.com/openscd/open-scd/tree/942-subscriber-add-lgos-and-lsvs) instead of main but don't mind which order things are done in, whatever is easy.

**Comment by danyill** [comment]
:cry: did you mean to close this @juancho0202?

**Comment by juancho0202** [comment]
@danyill not at all I'm so sorry, I thought it would automatically change to point to `main` instead of closing it 😔

**Comment by danyill** [comment]
OK, thanks @juancho0202.

I think it cannot repoint because you have force-pushed your branch and broken the dependency chain (and deleted the evidence!). I often force push too when my work is "private" but once there is a branch depending on it then to force push shared commits is to cause damage.

So I think Github is not providing any options on this PR because there is no way to reconnect it.

I'll do something when time allows.

**Pull request closed without merging**